orca-front50/src/main/groovy/com/netflix/spinnaker/orca/front50/DependentPipelineStarter.groovy

```diff
@@ -100,14 +100,18 @@ class DependentPipelineStarter implements ApplicationContextAware {
 
     def trigger = pipelineConfig.trigger
     //keep the trigger as the preprocessor removes it.
-    def expectedArtifacts = pipelineConfig.expectedArtifacts
 
     if (parentPipelineStageId != null) {
       pipelineConfig.receivedArtifacts = artifactResolver?.getArtifacts(parentPipeline.stageById(parentPipelineStageId))
     } else {
       pipelineConfig.receivedArtifacts = artifactResolver?.getAllArtifacts(parentPipeline)
     }
 
+    // For pipelines with dynamic source, it is required to perform the following steps:
+    //   1. Resolve source template artifact so the dynamic source expression can be evaluated.
```

Comment:
I'm not sure that I fully understand why we need to resolve artifacts twice here, and have some concerns about now always resolving artifacts twice.  I think partly it's that I don't fully understand what is a "pipeline with dynamic source" is.  Can you please give an example (or point to a test case) for what such a pipeline looks like and how it would get resolved?

Reply:
with pipeline with dynamic source I meant to have an expression to fetch source pipeline from artifacts such as 
```
"source": "{% for artifact in trigger.artifacts %}{% if artifact.name == 'pipeline-template' && artifact.type == 'http/file' %}http://{{artifact.reference}}{% endif %}{% endfor %}"
```
here a test with dynamic parent: https://github.com/spinnaker/orca/blob/2e07fcc7ecb5bf77e3a3a6a5df9e5d1be9c5eabd/orca-front50/src/test/groovy/com/netflix/spinnaker/orca/front50/DependentPipelineStarterSpec.groovy#L467

On why we need to resolve artifacts twice, the first time we need it in order to load parent pipeline, once we have loaded the parent pipeline, we need to resolve artifacts again as new artifacts could have been found in parent pipeline.